### DISCUSSION TITLE
Gas Mixing simulation failing due to FACTOR_OUTMEMORY

### DISCUSSION THREAD

**Discussion by gsgall** [idaholab/moose#25820] [Q&A General]
I am working on solving a single phase gas mixing simulation using FE-INS for the flow and modeling the gas mixing by modeling a mass fraction that varies spatially. 

In this simulation I am using an effective fluid density and viscosity which is a function of this mass fraction in the form 

rho = w * rho_1 + (1 - w) * rho_2

Where rho_1 is the density of one gas and rho_2 is the density of another. I am using a similar equation for the viscosity as well. 

I am running into an issue where my simulation is solving with out any real issues until a a certain point and then it just begins to crash, with the linear solve reporting `Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0 PC failed due to FACTOR_OUTMEMORY` and the non-linear solve subsequently reporting `Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0`. I don't believe that this really a memory issue, when looking at my memory usage there is a decent bit available when the solver starts to report `PC failed due to FACTOR_OUTMEMORY` and #23274 references that lu often reports this even when it is simply failing. 

 This failure occurs quite late into the simulation and there are no obvious issues when looking at the solution in Paraview. 

I looked at the techniques used to resolve the issues in #22425 but adding `off_diagonals_in_auto_scaling = true` only resulted in `Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0`occurring much faster than without it. I have also tried limiting the maximum time step taken with my adaptive time stepper but when the mixing profile develops to a similar point as without it I experience the same sort of crash. 

Input can be found here [https://github.com/gsgall/zapdos/blob/marias_jet/jet_fluid_only/mixing_helium_air.i](https://github.com/gsgall/zapdos/blob/marias_jet/jet_fluid_only/mixing_helium_air.i)

Below I have included the final output before my dt begins to start cutting until it fails  

```Steady-State Relative Differential Norm: 0.612157

Time Step 450, time = 0.0228275, dt = 0.0005

Performing automatic scaling calculation

 0 Nonlinear |R| = 2.246228e-03
      0 Linear |R| = 2.246228e-03
      1 Linear |R| = 2.767535e-08
      2 Linear |R| = 1.291543e-17
 1 Nonlinear |R| = 1.915105e-05
      0 Linear |R| = 1.915105e-05
      1 Linear |R| = 3.437917e-09
      2 Linear |R| = 4.502965e-20
 2 Nonlinear |R| = 3.476576e-05
      0 Linear |R| = 3.476576e-05
      1 Linear |R| = 5.903443e-16
 3 Nonlinear |R| = 1.042973e-04
      0 Linear |R| = 1.042973e-04
      1 Linear |R| = 2.562989e-22
 4 Nonlinear |R| = 1.129887e-04
      0 Linear |R| = 1.129887e-04
      1 Linear |R| = 1.537833e-23
 5 Nonlinear |R| = 1.260259e-04
      0 Linear |R| = 1.260259e-04
      1 Linear |R| = 1.238041e-24
 6 Nonlinear |R| = 5.649435e-05
      0 Linear |R| = 5.649435e-05
      1 Linear |R| = 3.097355e-26
 7 Nonlinear |R| = 8.256867e-05
      0 Linear |R| = 8.256867e-05
      1 Linear |R| = 3.666786e-20
 8 Nonlinear |R| = 7.822295e-05
      0 Linear |R| = 7.822295e-05
      1 Linear |R| = 1.532815e-26
 9 Nonlinear |R| = 3.042004e-05
      0 Linear |R| = 3.042004e-05
      1 Linear |R| = 1.030051e-26
10 Nonlinear |R| = 2.607432e-05
      0 Linear |R| = 2.607432e-05
      1 Linear |R| = 6.883952e-28
11 Nonlinear |R| = 2.607432e-05
      0 Linear |R| = 2.607432e-05
      1 Linear |R| = 7.123209e-27
12 Nonlinear |R| = 7.822295e-05
      0 Linear |R| = 7.822295e-05
      1 Linear |R| = 5.734877e-27
13 Nonlinear |R| = 5.214863e-05
      0 Linear |R| = 5.214863e-05
      1 Linear |R| = 4.477586e-27
14 Nonlinear |R| = 1.738288e-05
      0 Linear |R| = 1.738288e-05
      1 Linear |R| = 8.650959e-27
15 Nonlinear |R| = 8.691439e-06
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 15
 Solve Did NOT Converge!
  Finished Solving                                                                       [ 37.22 s] [  345 MB]
Aborting as solve did not converge

Solve failed, cutting timestep.

Time Step 450, time = 0.0225775, dt = 0.00025

Performing automatic scaling calculation

 0 Nonlinear |R| = 9.077309e-02
      0 Linear |R| = 9.077309e-02
      1 Linear |R| = 1.834745e-04
      2 Linear |R| = 1.741485e-16
 1 Nonlinear |R| = 1.815283e-01
      0 Linear |R| = 1.815283e-01
      1 Linear |R| = 1.917281e-09
 2 Nonlinear |R| = 2.463599e-01
      0 Linear |R| = 2.463599e-01
      1 Linear |R| = 4.083445e-16
 3 Nonlinear |R| = 2.074610e-01
      0 Linear |R| = 2.074610e-01
      1 Linear |R| = 1.019659e-20
 4 Nonlinear |R| = 1.555957e-01
      0 Linear |R| = 1.555957e-01
      1 Linear |R| = 8.905297e-22
 5 Nonlinear |R| = 4.408545e-01
      0 Linear |R| = 4.408545e-01
      1 Linear |R| = 1.957787e-16
 6 Nonlinear |R| = 3.630567e-01
      0 Linear |R| = 3.630567e-01
      1 Linear |R| = 5.307043e-24
 7 Nonlinear |R| = 9.602145e-16
 Solve Converged!
  Finished Solving                                                                       [ 21.03 s] [  343 MB]

Outlier Variable Residual Norms:
  w_he: 9.251459e-16

Postprocessor Values:
+----------------+-------------------+
| time           | pressure_integral |
+----------------+-------------------+
:                :                   :
|   1.582753e-02 |     -6.352747e-22 |
|   1.632753e-02 |     -2.964615e-21 |
|   1.682753e-02 |     -6.988022e-21 |
|   1.732753e-02 |     -6.352747e-22 |
|   1.782753e-02 |     -1.270549e-21 |
|   1.832753e-02 |     -2.329341e-21 |
|   1.882753e-02 |     -2.541099e-21 |
|   1.932753e-02 |      0.000000e+00 |
|   1.982753e-02 |      2.117582e-21 |
|   2.032753e-02 |     -8.470329e-22 |
|   2.082753e-02 |      4.658681e-21 |
|   2.132753e-02 |      2.964615e-21 |
|   2.182753e-02 |     -8.470329e-22 |
|   2.232753e-02 |     -1.482308e-21 |
|   2.257753e-02 |      0.000000e+00 |
+----------------+-------------------+


Scalar Variable Values:
+----------------+----------------+
| time           | lambda         |
+----------------+----------------+
:                :                :
|   1.582753e-02 |  -1.168477e-01 |
|   1.632753e-02 |  -1.168563e-01 |
|   1.682753e-02 |  -1.168650e-01 |
|   1.732753e-02 |  -1.168735e-01 |
|   1.782753e-02 |  -1.168819e-01 |
|   1.832753e-02 |  -1.168900e-01 |
|   1.882753e-02 |  -1.168978e-01 |
|   1.932753e-02 |  -1.169053e-01 |
|   1.982753e-02 |  -1.169124e-01 |
|   2.032753e-02 |  -1.169191e-01 |
|   2.082753e-02 |  -1.169253e-01 |
|   2.132753e-02 |  -1.169311e-01 |
|   2.182753e-02 |  -1.169363e-01 |
|   2.232753e-02 |  -1.169411e-01 |
|   2.257753e-02 |  -1.169426e-01 |
+----------------+----------------+

Steady-State Relative Differential Norm: 0.602031

Time Step 451, time = 0.0228275, dt = 0.00025

Performing automatic scaling calculation

 0 Nonlinear |R| = 1.262764e-03
      0 Linear |R| = 1.262764e-03
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_OUTMEMORY 
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
 Solve Did NOT Converge!
  Finished Solving                                                                       [  5.08 s] [  351 MB]
Aborting as solve did not converge

Solve failed, cutting timestep.

Time Step 451, time = 0.0227025, dt = 0.000125

Performing automatic scaling calculation
```

I have also included the solution of the gas fraction at the point when the solution has failed. 
![Screenshot from 2023-10-23 16-05-18](https://github.com/idaholab/moose/assets/66559200/051bde68-fd93-4af9-a177-888fe51d6d04)

**Comment by GiudGiud** [comment]
Hello

Indeed, `FACTOR_OUTMEMORY ` is often reported for  non-memory related reasons. 
Since your solver takes you quite far already, I would make sure you have a checkpoint before it crashes, then try to restart from there and tweak the solver options. 
I wonder if the fixed factor shift is enough for example. Or if mumps can succeed where superlu_dist crashes. 

Guillaume

**Reply by gsgall** [reply]
I have tried these suggestions with some positive results but my simulation is still not fully reaching the steady state. 

Increasing the fixed factor shift by an order of magnitude helped the simulation progress but not very significantly. Similar results were seen when changing the solve type to `mumps`.

Of all the things I have attempted the one that yielded the most progress was to change the order of the mass fraction variable from `SECOND` to `FIRST`. 

Currently I have stared a solve that restarts from the point where a previous solve failed but with an increase in the fixed factor shift. This has allowed the simulation to progress slightly more but the adaptive time stepper will periodically increase to about 1e-3 and then crash to 1e-12 and this seems to continue with the solution changing very slowly. I don't think this is because it close to the steady state. Based on some previous similar work I completed I have a solid idea of what the steady state solution should look like and this solve needs more time to evolve to the true steady state.

**Reply by GiudGiud** [reply]
Looks like you have some scalar advection by the INSAD velocities
@lindsayad is conservativeAdvection the right kernels for this? Should there be SUPG terms?

I am not a big fan of the average value constraint btw. I think it hurts the numerical solve.
I wonder if we can use the NSFVPressurePin, or a new NSPressure user object to achieve the same goal for FE Navier Stokes

**Reply by lindsayad** [reply]
In theory if the scalar transport is expect to be advection dominated, then you should be applying SUPG stabilization. The problem is that we only have readily available kernels for SUPG stabilization of energy advection, not for passive scalar advection. Would be a good contribution!

For the immediate issue I would see what happens with `-pc_factor_mat_solver_type superlu_dist` or `-pc_factor_mat_solver_type strumpack`

**Reply by gsgall** [reply]
I tried the strumpack option and had been using the superlu_dist but that didn't really make an impact. 

Based on what @GiudGiud said about the mean value zero method hurting the numerical solve I modified my fluid solve to use a pinned pressure node similar to the method in `modules/navier_stokes/test/tests/finite_element/ins/RZ_cone/rz_cone_by_parts_steady_nobcbc.i` this has resolved the issue I was having and I was able to achieve a steady state solve. 

@lindsayad Would you have any recommendations for literature on SUPG stabilization for passive scalar advection that I can look into?

**Reply by lindsayad** [reply]
You've had problems in the past with a pressure pin creating artifacts in the solution. Did you experience any of that here?

**Reply by GiudGiud** [reply]
I was pitching a postprcessing pressure pin btw. Constraint pressure pin is not my favorite either, but whatever works

**Reply by lindsayad** [reply]
There's not much special about passive scalar advection. Indeed, scalar advection was the first type of problem SUPG was developed for. I'm not finding the original paper on it, but [this](https://www.sciencedirect.com/science/article/abs/pii/0045782582900718) was probably one of the first few papers on SUPG (Thomas Hughes was the original creator of the method).

You could just follow what is done for temperature for the passive scalar, see the `INSADEnergy` kernels and the `INSAD3Eqn` and `INSADStabilized3Eqn` materials

**Reply by lindsayad** [reply]
> I was pitching a postprcessing pressure pin btw

The issue with this treatment is that in general a factorization-based preconditioner will not work with it (although exceptions may exist)

**Reply by gsgall** [reply]
I did end up running into some artifacts with the pressure pin again. The first case I ran with a pressure pin didn't show any artifacts in the solution. However, the second geometry I am working on, it's the same as the first but the distance from the tip of the outlet of the jet to the target is increased is now failing in the mixing solve due to similar artifacts. 

I will start to also take a look at NSFVPressurePin and try to apply something similar for my FE solve.

**Reply by gsgall** [reply]
I'll also take a look at that paper. Thank you for sharing!

**Reply by lindsayad** [reply]
Just be aware of what I said regarding a postprocessing (`NSFVPressurePin`-like process) to remove the nullspace... LU may not work

**Reply by gsgall** [reply]
I will keep that in mind thank you! 


I believe there is also a possible solution for this in a more indirect way as well that I would be interested in pursuing. The control volume I am interested in for this simulation is much smaller than the entire mesh being simulated. The instabilities that prevent my simulation from reaching steady state are quite from the control volume of interest and for the most part the control volume has reached steady state before instabilities begin to crop up. Is there a standard method in MOOSE for using steady state detection but in just a smaller part of the entire computational domain? 

Essentially I would like to create a small bounding box where steady state detection is used, is this possible?

**Reply by lindsayad** [reply]
No, this is not currently possible. But we are planning to add a custom convergence criteria system in the next fiscal year, so you could imagine writing a convergence criteria class that does what you're requesting

**Reply by gsgall** [reply]
Sounds great! Thank you!
